### PR TITLE
Model submission for marble_husky sensor configuration 6 by MARBLE

### PR DESCRIPTION
**Pull request by bfotheri**
This sensor configuration represents the vehicle MARBLE is presently using and plans to use for the final circuit event this fall. It differs from previous sensor configurations in that it now has 4 RGB cameras facing orthogonal directions, a vertical OUSTER LIDAR, and the front two D435i's have been replaced with Pico Flexx cameras. 

I need to discuss with my team but apparently validation data for this and previous configurations were submitted to OSRF in the fall of 2020 by our teammate Kyle Harlow on behalf of MARBLE. I will ask him where that is, but on your end could you also check if that was received? Thank you guys!

### CODE REVIEW

**Comment by angelacmaio** [comment]
Can you upload the validation data to the DARPA Box link that has since been provided in the [Model Preparation Guide](https://subtchallenge.com/resources/Simulation_Model_Preparation_Guide.pdf)?

**Comment by bfotheri** [comment]
Validation data has been successfully uploaded to the link provided in the Model Preparation Guide. File is named "MARBLE_HUSKY.zip"

**Comment by bfotheri** [comment]
Also, MARBLE is currently using this vehicle as a marsupial pair in preparation for the final circuit systems event. I have local changes to competition.ign that allow this locally, but was unsure of the procedure for merging those in. Should I add those marsupial changes and commit them for review or should some verification/discussion happen prior?

**Review by acschang — CHANGES_REQUESTED** [review]
The initial assessment of this sensor configuration submission is complete, however there are some issues to resolve before this sensor configuration can be merged and utilized in a competitive setting:

- [ ] Please update the Ouster LIDAR and Pico Flexx sensor parameters per the in-line comments.
- [ ] Please verify the mass listed in the `specifications.md` file matches the total mass in the `model.sdf`.
- [ ] Please add the `wheel_slip` plugin section to the `spawner.rb` file.
- [ ] Please clarify if the sensor suite cost for sensor config 6 matches the value in the `specifications.md` file.
- [ ] Please update the sensor specifications link for the FLIR camera in the `specifications.md` file.
- [x] Please add all missing components to the RViz visualization.
- [ ] Please update `vehicle_topics.launch` file in sensor config 3 to include transports for the thermal camera and the `breadcrumbs_remaining` transport for compatibility with sensor config 4.

Please address the issues noted above and add commits to this pull request as soon as possible. The submissions are expected to be incorporated into the SubT Virtual Testbed

File: submitted_models/marble_husky_sensor_config_6/model.sdf [hunk 1]
```diff
@@ -0,0 +1,1246 @@
...
+                <samples>1024</samples>
+                <resolution>1</resolution>
+                <min_angle>-3.14159</min_angle>
+                <max_angle>3.14159</max_angle>
+              </horizontal>
+              <vertical>
+                <samples>64</samples>
+                <resolution>1</resolution>
+                <min_angle>-0.2897</min_angle>
+                <max_angle>0.2897</max_angle>
```
[the remark this answers is not in the document]
Comment (dan-riley) [reply submitted_models/marble_husky_sensor_config_6/model.sdf:630]: > The OS1-64's vertical field of view is +-22.5 degrees. Please update the `min_angle` and `max_angle` parameters here to: `0.392699`.

We have Gen 1 OS-1's, which only have a 33 degree vfov.

File: submitted_models/marble_husky_sensor_config_6/model.sdf [hunk 2]
```diff
@@ -0,0 +1,1246 @@
...
```
[the remark this answers is not in the document]
Comment (dan-riley) [reply submitted_models/marble_husky_sensor_config_6/model.sdf:595]: We have Gen 1 OS-1's, which only have a 33 degree vfov.

File: submitted_models/marble_husky_sensor_config_6/model.sdf [hunk 3]
```diff
@@ -0,0 +1,1246 @@
...
+            </scan>
+            <range>
+              <min>0.8</min>
+              <max>120</max>
+              <resolution>0.003</resolution>
+            </range>
+            <noise>
+              <type>gaussian</type>
+              <mean>0</mean>
+              <stddev>0.015</stddev>
```
[the remark this answers is not in the document]
Comment (bfotheri) [reply submitted_models/marble_husky_sensor_config_6/model.sdf:641]: Value appears to already be 0.015, please clarify

File: submitted_models/marble_husky_sensor_config_6/model.sdf [hunk 4]
```diff
@@ -0,0 +1,1246 @@
...
+                <samples>64</samples>
+                <resolution>1</resolution>
+                <min_angle>-0.2897</min_angle>
+                <max_angle>0.2897</max_angle>
+              </vertical>
+            </scan>
+            <range>
+              <min>0.8</min>
+              <max>120</max>
+              <resolution>0.003</resolution>
```
[the remark this answers is not in the document]
Comment (bfotheri) [reply submitted_models/marble_husky_sensor_config_6/model.sdf:636]: Updated, will push soon

File: submitted_models/marble_husky_sensor_config_6/model.sdf [hunk 5]
```diff
@@ -0,0 +1,1246 @@
...
+            </range>
+            <noise>
+              <type>gaussian</type>
+              <mean>0</mean>
+              <stddev>0.015</stddev>
```
[the remark this answers is not in the document]
Comment (bfotheri) [reply submitted_models/marble_husky_sensor_config_6/model.sdf:606]: Value appears to already be 0.015, please clarify

File: submitted_models/marble_husky_sensor_config_6/model.sdf [hunk 6]
```diff
@@ -0,0 +1,1246 @@
...
+                <samples>64</samples>
+                <resolution>1</resolution>
+                <min_angle>-0.2897</min_angle>
+                <max_angle>0.2897</max_angle>
+              </vertical>
+            </scan>
+            <range>
+              <min>0.8</min>
+              <max>120</max>
+              <resolution>0.003</resolution>
```
[the remark this answers is not in the document]
Comment (bfotheri) [reply submitted_models/marble_husky_sensor_config_6/model.sdf:601]: Updated, will push soon

File: submitted_models/marble_husky_sensor_config_6/model.sdf [hunk 7]
```diff
@@ -0,0 +1,1246 @@
...
+      <sensor name="tof_right" type="depth_camera">
+        <pose frame="">0.527 -0.015 0.429 -1.7970717 0.4763404 -0.4653087  </pose>
+        <always_on>1</always_on>
+        <update_rate>10</update_rate>
+        <camera name="tof_right">
+            <horizontal_fov>1.0821</horizontal_fov>
+            <lens>
+                <intrinsics>
+                    <!-- fx = fy = width / ( 2 * tan (hfov / 2 ) ) -->
+                    <fx>59.55</fx>
```
[the remark this answers is not in the document]
Comment (bfotheri) [reply submitted_models/marble_husky_sensor_config_6/model.sdf:331]: That's correct, updated, will push soon.

File: submitted_models/marble_husky_sensor_config_6/model.sdf [hunk 8]
```diff
@@ -0,0 +1,1246 @@
...
+        <pose frame="">0.527 0.015 0.429 -1.344521 0.4763404 0.4653087  </pose>
+        <!-- <pose>0.0891 0.095 0.015 0.0 -0.5236 0.41015</pose> -->
+        <always_on>1</always_on>
+        <update_rate>10</update_rate>
+        <camera name="tof_left">
+            <horizontal_fov>1.0821</horizontal_fov>
+            <lens>
+                <intrinsics>
+                    <!-- fx = fy = width / ( 2 * tan (hfov / 2 ) ) -->
+                    <fx>59.55</fx>
```
[the remark this answers is not in the document]
Comment (bfotheri) [reply submitted_models/marble_husky_sensor_config_6/model.sdf:258]: That's correct, updated, will push soon.

File: submitted_models/marble_husky_sensor_config_6/specifications.md [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+* Vehicle Links:
+  * https://clearpathrobotics.com/husky-unmanned-ground-vehicle-robot/
+
+* Sensor specification links:
+  * FLIR Backfly Camera - https://www.edmundoptics.com/p/bfly-pge-05s2c-cs-13-blackflyreg-poe-gige-color-camera/30315/
+  * Ouster 3D Lidar (64 Channel) - https://ouster.com/products/os1-lidar-sensor/
+  * RPLidar S1 Planar Lidar - https://www.slamtec.com/en/Lidar/S1Spec
+  * Picco Flexx Camera - https://www.automation24.com/development-kit-pmd-vision-r-camboard-pico-flexx-700-000-094
+  * IMU: Microstrain 3DM-GX5-25 - datasheet: https://www.microstrain.com/sites/default/files/applications/files/3dm-gx5-25_datasheet_8400-0093_rev_n.pdf
+    * Explanation of sensor parameter derivations:
```
[the remark this answers is not in the document]
Comment (bfotheri) [reply submitted_models/marble_husky_sensor_config_6/specifications.md:61]: Updated, will push soon. Validation data was uploaded to the following [location](https://github.com/osrf/subt/pull/850#issuecomment-810305095) referenced in the comment , for the time being but is not in youtube or google drive format yet

File: submitted_models/marble_husky_sensor_config_6/specifications.md [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+## Usage Instructions
+The robot can be used in the same manner as the COSTAR Husky robot.  It accepts twist messages on the cmd_vel topic.  
+
+## Usage Rights
+The same Rights are granted for the configuration as for the COSTAR Husky. No additional restrictions have to be taken into account for this configuration.
+
+### Cost and Scale
+The MARBLE Husky has the following estimate commercial cost components:
+* Base vehicle: $25,000
+* Sensor suite: $12,000
```
[the remark this answers is not in the document]
Comment (bfotheri) [reply submitted_models/marble_husky_sensor_config_6/specifications.md:18]: Updated, will push soon

File: submitted_models/marble_husky_sensor_config_3/launch/vehicle_topics.launch [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+<launch>
+  <arg name="world_name"/>
+  <arg name="name"/>
+  <arg name="link_name" value="base_link"/>
+  <!--<arg name="pan_joint_name" value="pan_gimbal_joint" />
+  <arg name="tilt_joint_name" value="tilt_gimbal_joint"/>-->
+  <arg name="sensor_prefix" value="/world/$(arg world_name)/model/$(arg name)/link/$(arg link_name)/sensor"/>
+  <arg name="gimbal_sensor_prefix" value="/world/$(arg world_name)/model/$(arg name)/link/tilt_gimbal_link/sensor"/>
+  <arg name="breadcrumbs" default="1"/>
+  <group ns="$(arg name)">
```
[the remark this answers is not in the document]
Comment (bfotheri) [reply submitted_models/marble_husky_sensor_config_3/launch/vehicle_topics.launch:11]: Updated, will push soon

File: submitted_models/marble_husky_sensor_config_3/launch/vehicle_topics.launch [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+
+    <group if="$(arg breadcrumbs)">
+      <node
+        pkg="ros_ign_bridge"
+        type="parameter_bridge"
+        name="ros_ign_bridge_breadcrumbs"
+        args="/model/$(arg name)/breadcrumb/deploy@std_msgs/Empty]ignition.msgs.Empty">
+        <remap from="/model/$(arg name)/breadcrumb/deploy" to="breadcrumb/deploy"/>
+      </node>
+    </group>
```
[the remark this answers is not in the document]
Comment (bfotheri) [reply submitted_models/marble_husky_sensor_config_3/launch/vehicle_topics.launch:268]: Updated, will push soon

File: submitted_models/marble_husky_sensor_config_3/launch/vehicle_topics.launch [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+
+    <!-- Gas sensor -->
+    <node
+      pkg="ros_ign_bridge"
+      type="parameter_bridge"
+      name="ros_ign_bridge_gas"
+      args="/model/$(arg name)/gas_detected@std_msgs/Bool[ignition.msgs.Boolean">
+      <remap from="/model/$(arg name)/gas_detected" to="gas_detected"/>
+    </node>
+
```
[the remark this answers is not in the document]
Comment (bfotheri) [reply submitted_models/marble_husky_sensor_config_3/launch/vehicle_topics.launch:259]: updated, will push soon

File: submitted_models/marble_husky_sensor_config_3/.vscode/c_cpp_properties.json [hunk 14]
```diff
@@ -0,0 +1,20 @@
+{
```
[the remark this answers is not in the document]
Comment (acschang) [reply submitted_models/marble_husky_sensor_config_3/.vscode/c_cpp_properties.json]: Fixed in https://github.com/osrf/subt/pull/850/commits/a21c326a0059be38dd5920bf592e2b6e2d5af2d8

File: submitted_models/marble_husky_sensor_config_3/.vscode/settings.json [hunk 15]
```diff
@@ -0,0 +1,5 @@
+{
```
[the remark this answers is not in the document]
Comment (acschang) [reply submitted_models/marble_husky_sensor_config_3/.vscode/settings.json]: Fixed in https://github.com/osrf/subt/pull/850/commits/a21c326a0059be38dd5920bf592e2b6e2d5af2d8

File: submitted_models/marble_husky_sensor_config_6/specifications.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (acschang) [reply submitted_models/marble_husky_sensor_config_6/specifications.md]: Fixed in https://github.com/osrf/subt/pull/850/commits/a21c326a0059be38dd5920bf592e2b6e2d5af2d8

File: submitted_models/marble_husky_sensor_config_6/specifications.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (acschang) [reply submitted_models/marble_husky_sensor_config_6/specifications.md]: Fixed in https://github.com/osrf/subt/pull/850/commits/a21c326a0059be38dd5920bf592e2b6e2d5af2d8

File: submitted_models/marble_husky_sensor_config_6/specifications.md [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+
+### Diversions from Physical Hardware of MARBLE Husky
+Computers were installed in the payload area of the husky and these have been roughly modeled (the rail and computer bay are shown in the model.sdf model).  The MARBLE vehicle uses an AMD Ryzen processor (32-core) with 64 GB of RAM.  It has a cooling system as well.  
+
+## X4 Validation and Specification Links
+* Vehicle Links:
+  * https://clearpathrobotics.com/husky-unmanned-ground-vehicle-robot/
+
+* Sensor specification links:
+  * FLIR Backfly Camera - https://www.edmundoptics.com/p/bfly-pge-05s2c-cs-13-blackflyreg-poe-gige-color-camera/30315/
```
[the remark this answers is not in the document]
Comment (acschang) [reply submitted_models/marble_husky_sensor_config_6/specifications.md]: Fixed in https://github.com/osrf/subt/pull/850/commits/a21c326a0059be38dd5920bf592e2b6e2d5af2d8

File: submitted_models/marble_husky_sensor_config_6/model.sdf (hunk 7 above)
[the remark this answers is not in the document]
Comment (acschang) [reply submitted_models/marble_husky_sensor_config_6/model.sdf]: Fixed in https://github.com/osrf/subt/pull/850/commits/a21c326a0059be38dd5920bf592e2b6e2d5af2d8

File: submitted_models/marble_husky_sensor_config_6/model.sdf (hunk 2 above)
[the remark this answers is not in the document]
Comment (acschang) [reply submitted_models/marble_husky_sensor_config_6/model.sdf:595]: The [provided documentation](https://ouster.com/products/os1-lidar-sensor/) indicates the VFoV should be +-22.5 degrees.  Is the included documentation incorrect?
Screenshot from 2021-05-18 15-35-48

File: submitted_models/marble_husky_sensor_config_6/model.sdf (hunk 6 above)
[the remark this answers is not in the document]
Comment (acschang) [reply submitted_models/marble_husky_sensor_config_6/model.sdf]: Fixed in https://github.com/osrf/subt/pull/850/commits/a21c326a0059be38dd5920bf592e2b6e2d5af2d8

File: submitted_models/marble_husky_sensor_config_6/model.sdf (hunk 4 above)
[the remark this answers is not in the document]
Comment (acschang) [reply submitted_models/marble_husky_sensor_config_6/model.sdf]: Fixed in https://github.com/osrf/subt/pull/850/commits/a21c326a0059be38dd5920bf592e2b6e2d5af2d8

File: submitted_models/marble_husky_sensor_config_6/model.sdf (hunk 8 above)
[the remark this answers is not in the document]
Comment (acschang) [reply submitted_models/marble_husky_sensor_config_6/model.sdf:258]: Fixed in https://github.com/osrf/subt/pull/850/commits/a21c326a0059be38dd5920bf592e2b6e2d5af2d8

File: submitted_models/marble_husky_sensor_config_3/launch/vehicle_topics.launch (hunk 11 above)
[the remark this answers is not in the document]
Comment (acschang) [reply submitted_models/marble_husky_sensor_config_3/launch/vehicle_topics.launch:12]: Added in https://github.com/osrf/subt/pull/850/commits/a21c326a0059be38dd5920bf592e2b6e2d5af2d8

File: submitted_models/marble_husky_sensor_config_3/launch/vehicle_topics.launch (hunk 12 above)
[the remark this answers is not in the document]
Comment (acschang) [reply submitted_models/marble_husky_sensor_config_3/launch/vehicle_topics.launch:306]: Added in https://github.com/osrf/subt/pull/850/commits/a21c326a0059be38dd5920bf592e2b6e2d5af2d8

File: submitted_models/marble_husky_sensor_config_6/model.sdf (hunk 2 above)
[the remark this answers is not in the document]
Comment (acschang) [reply submitted_models/marble_husky_sensor_config_6/model.sdf:595]: So it appears that the Ouster website is displaying Gen 2 sensor parameters by default and Gen 1's parameters are listed [elsewhere](https://data.ouster.io/downloads/datasheets/datasheet-gen1-v2p0-os1.pdf).  I'll update the link to the documentation in this case however the documentation links are definitely deprecated for other Ouster Gen 1 sensors in the repository now.

**Comment by acschang** [comment]
Continued in #918

File: submitted_models/marble_husky_sensor_config_6/model.sdf (hunk 2 above)
[the remark this answers is not in the document]
Comment (acschang) [reply submitted_models/marble_husky_sensor_config_6/model.sdf:595]: Addressed in https://github.com/osrf/subt/pull/918/commits/fff0990fcc17a5ea93d5a73a87bb58bf88228b7c

File: submitted_models/marble_husky_sensor_config_6/model.sdf (hunk 1 above)
[the remark this answers is not in the document]
Comment (acschang) [reply submitted_models/marble_husky_sensor_config_6/model.sdf:630]: Addressed in https://github.com/osrf/subt/pull/918/commits/fff0990fcc17a5ea93d5a73a87bb58bf88228b7c based upon https://github.com/osrf/subt/pull/850#discussion_r624578255

File: submitted_models/marble_husky_sensor_config_6/launch/spawner.rb [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+        <start_on_motion>true</start_on_motion>
+      </plugin>
+      <!-- Gas Sensor plugin -->"
+      <plugin filename="libGasEmitterDetectorPlugin.so"
+        name="subt::GasDetector">
+        <topic>/model/#{_name}/gas_detected</topic>
+        <update_rate>10</update_rate>
+        <type>gas</type>
+      </plugin>
+      <plugin filename="libignition-gazebo-breadcrumbs-system.so"
```
[the remark this answers is not in the document]
Comment (acschang) [reply submitted_models/marble_husky_sensor_config_6/launch/spawner.rb:61]: Addressed in https://github.com/osrf/subt/pull/850/commits/a21c326a0059be38dd5920bf592e2b6e2d5af2d8

File: submitted_models/marble_husky_sensor_config_6/model.sdf (hunk 3 above)
[the remark this answers is not in the document]
Comment (acschang) [reply submitted_models/marble_husky_sensor_config_6/model.sdf:641]: I meant to type `0.01` to meet the repository standard however I made a typo.

Fixed in https://github.com/osrf/subt/pull/918/commits/c540270ac30d3bfbda3ab2f00b5ad217bf5b4abe

File: submitted_models/marble_husky_sensor_config_6/model.sdf (hunk 5 above)
[the remark this answers is not in the document]
Comment (acschang) [reply submitted_models/marble_husky_sensor_config_6/model.sdf:606]: Fixed in https://github.com/osrf/subt/pull/918/commits/c540270ac30d3bfbda3ab2f00b5ad217bf5b4abe

File: submitted_models/marble_husky_sensor_config_3/model.config [hunk 18]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0"?>
+<model>
+  <name>marble_husky_sensor_config_3</name>
+  <sdf version="1.6">model.sdf</sdf>
+  <description>
+     CU's husky config with 3 D435i RGBD cameras, a pan-tilt turret, the Ouster OS1 64 channel lidar,
```
[the remark this answers is not in the document]
Comment (acschang) [reply submitted_models/marble_husky_sensor_config_3/model.config]: Fixed in https://github.com/osrf/subt/pull/918/commits/a21c326a0059be38dd5920bf592e2b6e2d5af2d8

**Pull request closed without merging**